Generate futures-based bluez interface

alsuren opened the pull request:
This is now running on the raspberry pi, and seems to be doing ok

TODO:

- [x] check the license of blurz and see if it's acceptable to copy-pasta the BluetoothEvent enum across (or whether I need to strip it out)
- [x] ~have a go at generating interfaces from the khadas VIM3 and see if they differ~
- [x] ~have a go at *running* on the VIM3 with a set of named sensors~
- [x] have a look over the code together and decide what wants reshaping

alsuren committed:
ssh pi@raspberrypi.local .cargo/bin/dbus-codegen-rust --system-bus --destination org.bluez --path /org/bluez/hci0/dev_A4_C1_38_1E_0A_E8 > mod.rs

qwandor committed:
Disconnect if starting notifications failed.

This should hopefully leave things in a cleaner state to retry later.

alsuren committed:
dbus-codegen-rust --futures

alsuren committed:
bluez-generated will be a lib

alsuren committed:
nonblock client only

alsuren committed:
just a single interface

alsuren committed:
script to introspect bluez recursively

alsuren committed:
generated specs

qwandor committed:
Split out separate loops

alsuren committed:
format xml and strip node names

alsuren committed:
formatted xml

alsuren committed:
shfmt -w -bn -i 4 **/*.sh

alsuren committed:
shfmt -w -i 4 -bn -sr  **/*.sh

alsuren committed:
put shfmt config in to .editorconfig

alsuren committed:
generate rust interfaces

alsuren committed:
generated interfaces

alsuren committed:
move generated files into src

alsuren committed:
use dbus-codegen-rust from master

alsuren committed:
got it compiling

alsuren committed:
Delete read-all-devices. It is dead weight.

alsuren committed:
Merge remote-tracking branch 'alsuren/tasks' into bluez-generated

alsuren committed:
copy-pasta BluetoothEvent from blurz

alsuren committed:
use dbus_tokio to listen to signal stream

alsuren committed:
optional SUFFIX for run.sh to avoid stomping prod

alsuren committed:
remove unused deps spotted by cargo-udeps

alsuren committed:
slow down the connect loop

alsuren committed:
print based debugging

alsuren committed:
Revert "Delete read-all-devices. It is dead weight."

This reverts commit 16980475a113b32299d956fe211e0f191ea58d96.

alsuren committed:
minimal event printer

alsuren committed:
rsync and run a single binary

alsuren committed:
it's called the system bus

alsuren committed:
print decoded readings

alsuren committed:
turn the adapter on

alsuren committed:
list all devices by reading GetManagedObjects data

alsuren committed:
kill off dbg!()s

alsuren committed:
get_sensors() -> Vec<Sensor>

alsuren committed:
WIP connect_start_sensor()

alsuren committed:
30 second dbus timeout everywhere

alsuren committed:
check ServiceData's keys rather than UUIDs

alsuren committed:
turn device_path back into a String

alsuren committed:
WIP on porting publish-mqtt

alsuren committed:
omg no warnings

alsuren committed:
remove blurz

alsuren committed:
factor out check_for_sensors()

alsuren committed:
port to anyhow

alsuren committed:
poor man's backtraces

alsuren committed:
survive when discovery is already started

alsuren committed:
a bit less spam

alsuren committed:
leave sensor connected for a bit if subscribing fails

alsuren committed:
kill off competing processes in run.sh

alsuren commented on bluez-generated/introspect.sh:
```diff
@@ -0,0 +1,63 @@
+
+if [ ${INTROSPECT:-1} = 1 ]; then
+    ssh pi@raspberrypi.local \
+        gdbus introspect --system --dest org.bluez --object-path / --recurse \
```
- [x] Split this out into a $GDBUS so that it can be set to 'gdbus' on localhost or 'sshpi@raspberrypi.local gdbus' for remote introspection.

alsuren commented on bluez-generated/src/bluetooth_event.rs:
```diff
@@ -0,0 +1,113 @@
+use dbus::{arg::cast, arg::RefArg, arg::TypeMismatchError, arg::Variant, Message};
```
- [x] add a copyright header here saying where it was copied from

alsuren commented on bluez-generated/src/generated/adapter1.rs:
```diff
@@ -0,0 +1,132 @@
+
+pub trait OrgBluezAdapter1 {
+    fn start_discovery(&self) -> nonblock::MethodReply<()>;
+    fn set_discovery_filter(&self, properties: ::std::collections::HashMap<&str, arg::Variant<Box<dyn arg::RefArg>>>) -> nonblock::MethodReply<()>;
```
- [x] run these files through rustfmt?

alsuren replied:
🤷 seems like they're fine

alsuren commented on bluez-generated/src/lib.rs:
```diff
@@ -0,0 +1,2 @@
+pub mod bluetooth_event;
+pub mod generated;
```
- [x] think about re-exporting everything from the top level?

alsuren commented on homie/src/lib.rs:
```diff
@@ -164,13 +163,8 @@ impl HomieDeviceBuilder {
-        ),
-        SendError<Request>,
-    > {
+    ) -> Result<(HomieDevice, impl Future<Output = Result<(), anyhow::Error>>), SendError<Request>>
```
- [ ] revert the homie anyhow port for now

alsuren commented on mijia/src/lib.rs:
```diff
@@ -1,137 +1,104 @@
-        );
-        if let Ok(service_data) = device.get_service_data() {
-            println!("Service data: {:?}", service_data);
+pub async fn get_sensors(
```
- [ ] rename this to find_sensors()? seems similar enough.

alsuren commented on mijia/src/lib.rs:
```diff
@@ -1,137 +1,104 @@
+        .into_iter()
+        .filter_map(|(path, interfaces)| {
+            // FIXME: can we generate a strongly typed deserialiser for this,
+            // based on the introspection data?
```
- [x] add a ticket for this when we make a new repo (what should the new repo be called?)

alsuren commented on publish-mqtt/Cargo.toml:
```diff
@@ -17,3 +16,8 @@ rustls = "0.17"
 tokio = "0.2.22"
 dotenv = "0.15.0"
+bluez-generated = { path = "../bluez-generated" }
+dbus-tokio = "*"
```
- [x] turn all of these * dependencies into concrete versions, and sort dependencies alphabetically

alsuren commented on publish-mqtt/src/main.rs:
```diff
@@ -187,77 +191,150 @@ impl Sensor {
             )
-            .await?;
+            .await
+            .with_context(|| std::line!().to_string())?;
```
- [ ] this was really useful for testing, but maybe we can use stable-eyre and set RUST_LIB_BACKTRACE=1 to get the same benefit for less typing?

alsuren commented on publish-mqtt/src/main.rs:
```diff
@@ -187,77 +191,150 @@ impl Sensor {
+        .await
+        .with_context(|| std::line!().to_string())?;
+
+    let state = Arc::new(Mutex::new(SensorState {
```
- [ ] Think about renaming our structs and enums

we currently have:
* SensorState - contains a Vec and a VecDeque and homie
* Sensor - various props/state about a single sensor
* SensorProps - the mac address and dbus path
* ConnectionStatus - what stage of connecting we're in. Currently duplicates some of the Vec/VecDequeue split from SensorState.

alsuren commented on publish-mqtt/src/main.rs:
```diff
@@ -187,77 +191,150 @@ impl Sensor {
+    let t1 = async {
+        loop {
+            let now = Instant::now();
+            if now > next_scan_due {
```
```suggestion
            if now > next_scan_due && !state.have_all_named_sensors() {
```
 or something?

alsuren commented on publish-mqtt/src/main.rs:
```diff
@@ -187,77 +191,150 @@ impl Sensor {
+    adapter
+        .start_discovery()
+        .await
+        .unwrap_or_else(|err| println!("starting discovery failed {:?}", err));
```
- [ ] make this error println sound less scary (we expect it to explode every time other than the first time we call it)

alsuren commented on publish-mqtt/src/main.rs:
```diff
@@ -187,77 +191,150 @@ impl Sensor {
+    let state = &mut *state.lock().await;
+    for props in sensors {
+        // Race Condition: When connecting, we remove from
+        // sensors_to_connect before adding to sensors_connected
```
eventually I will probably refactor the pair of lists into a single store that uses ConnectionStatus as the source of truth, and then this race condition will go away.

qwandor-google replied:
This isn't actually a race condition, is it? As we you hold the state lock the whole time, so nobody else can touch these two collections in the meantime.

alsuren commented on publish-mqtt/src/main.rs:
```diff
@@ -90,29 +73,45 @@ async fn main() -> Result<(), Box<dyn Error + Send + Sync>> {
+#[derive(Debug)]
+enum ConnectionStatus {
+    Unknown,
+    SubscribingFailedOnce,
```
- [x] add docstrings to each of these variants. This is the only variant that's really interesting at the moment (see `connect_start_sensor()` for details)

alsuren commented on publish-mqtt/src/main.rs:
```diff
@@ -303,44 +407,61 @@ async fn disconnect_first_stale_sensor(
+    println!("Subscribing to events");
+    let mut rule = dbus::message::MatchRule::new();
+    rule.msg_type = Some(dbus::message::MessageType::Signal);
+    rule.sender = Some(dbus::strings::BusName::new("org.bluez").map_err(|s| anyhow::anyhow!(s))?);
```
```suggestion
    rule.sender = Some(dbus::strings::BusName::new("org.bluez").map_err(anyhow::Err::new);
```

alsuren commented on publish-mqtt/src/main.rs:
```diff
@@ -90,29 +73,45 @@ async fn main() -> Result<(), Box<dyn Error + Send + Sync>> {
+    let (dbus_resource, dbus_conn) = dbus_tokio::connection::new_system_sync()?;
+    let dbus_handle = tokio::spawn(async {
+        let err = dbus_resource.await;
+        Err::<(), anyhow::Error>(anyhow::anyhow!(err))
```
```suggestion
        Err::<(), anyhow::Error>(anyhow::Err::new(err))
```

alsuren commented on publish-mqtt/src/main.rs:
```diff
@@ -303,44 +407,61 @@ async fn disconnect_first_stale_sensor(
         BluetoothEvent::Value { object_path, value } => {
             // TODO: Make this less hacky.
-            let device_path = match object_path.strip_suffix(SERVICE_CHARACTERISTIC_PATH) {
+            let object_path = match object_path.strip_suffix(SERVICE_CHARACTERISTIC_PATH) {
```
- [x] I prefer the name `Sensor.object_path` to mean "the object path for the root of the sensor", but in this specific function let's switch it back to device_path.

alsuren commented on read-all-devices/src/main.rs:
```diff
@@ -1,53 +1,208 @@
-    thread::sleep(Duration::from_millis(5000));
-    for device in &connected_sensors {
-        explore_device::explore_gatt_profile(bt_session, &device);
+#[tokio::main(core_threads = 3)]
```
- [x] we can get away with letting tokio decide how many cores it wants now, because nothing is blocking for more than a fraction of a second.

alsuren commented on read-all-devices/src/main.rs:
```diff
@@ -1,53 +1,208 @@
+    res?;
+    Ok(())
+}
+async fn connect_start_sensor<'a>(
```
- [x] ... yeah, there's quite a lot of copy-paste now :/

alsuren commented on read-all-devices/src/main.rs:
```diff
@@ -1,53 +1,208 @@
+}
+
+#[derive(Debug)]
+struct Sensor {
```
This could definitely be shared between read-all-devices and publish-mqtt.

alsuren commented on read-all-devices/src/main.rs:
```diff
@@ -1,53 +1,208 @@
+    last_update_timestamp: Instant,
+}
+
+async fn get_sensors(
```
- [x] 📋 🍝

alsuren commented on read-all-devices/src/main.rs:
```diff
@@ -1,53 +1,208 @@
+    Ok(paths)
+}
+
+async fn service_bluetooth_event_queue(
```
- [x] 📋 🍝

alsuren commented on bluez-generated/introspect.sh:
```diff
@@ -0,0 +1,63 @@
+# Bash >= 4 does not come with osx because of its GPLv3 license.
+# Install it via homebrew to get associative array support.
+
+set -euxo pipefail
```
comments 
- [x] prerequisites
- [x] what it's actually doing

alsuren commented on bluez-generated/introspect.sh:
```diff
@@ -0,0 +1,63 @@
+            --client=nonblock \
+            --methodtype=none \
+            > src/generated/$modname.rs
+        echo "mod $modname;" >> src/generated/mod.rs
```
- [x] I might have hand-edited this. Re-run and make sure it doesn't break things.

alsuren commented on mijia/src/lib.rs:
```diff
@@ -1,137 +1,104 @@
+                .as_iter()?
+                .filter_map(|addr| addr.as_str())
+                .next()?
+                .to_string();
```
- [ ] `VariantExt::as_string() -> Option<String>`

alsuren replied:
I'm going to do this later. The upstream maintainer says he has pushed a commit which might help with this.

alsuren commented on mijia/src/lib.rs:
```diff
@@ -1,137 +1,104 @@
+                // InternalDict(...)
+                .as_iter()?
+                .tuples::<(_, _)>()
+                .map(|(k, v)| (k.as_str().unwrap().into(), v))
```
theoretically this unwrap could fail.

alsuren commented on mijia/src/lib.rs:
```diff
@@ -1,137 +1,104 @@
-    }
+        })
+        .collect();
+    Ok(paths)
```
- [x] props or something?

alsuren commented on publish-mqtt/src/main.rs:
```diff
@@ -90,29 +73,45 @@ async fn main() -> Result<(), Box<dyn Error + Send + Sync>> {
-        bluetooth_mainloop(homie).await.unwrap();
+    // Connect to the D-Bus system bus (this is blocking, unfortunately).
+    let (dbus_resource, dbus_conn) = dbus_tokio::connection::new_system_sync()?;
+    let dbus_handle = tokio::spawn(async {
```
```suggestion
    // The resource is a task that should be spawned onto a tokio compatible
    // reactor ASAP. If the resource ever finishes, you lost connection to D-Bus.
    let dbus_handle = tokio::spawn(async {
```

alsuren commented on publish-mqtt/src/main.rs:
```diff
@@ -272,17 +349,44 @@ async fn connect_first_sensor_in_queue(
+        }
+        Err(e) => {
+            // If starting notifications failed, disconnect so that we start again from a clean
+            // state next time.
```
- [x] make a task to drive this off the services_resolved property changed signal

alsuren committed:
mijia::session::BluetoothSession::new()

alsuren committed:
MijiaSession::event_stream()

alsuren committed:
connect and disconnect

alsuren committed:
cargo upgrade and sort deps

alsuren committed:
kill off read-all-devices

It will eventually be split into:
* a no-mqtt mode for publish-mqtt.
* an example in mijia crate that takes a bluetooth address as env var.
* an example in mijia crate that continuously scans and connects to
  each sensor in turn for a configurable timeout/reading-count.

alsuren committed:
let tokio decide how many threads to use

alsuren committed:
Revert "poor man's backtraces"

This reverts commit eb6fd76ae0b9c1b738f94d537b8644f48bcac5ab.

alsuren committed:
Revert "port to anyhow"

This reverts commit 5498129f0417f1464aa23c99fc2a82fe87e62e4d.

alsuren committed:
remove anyhow from homie again

alsuren committed:
introspect.sh fixes

alsuren committed:
stfu; bluez only

alsuren committed:
BluetoothEvent copyright header

alsuren committed:
export everything from the same module

alsuren committed:
fix error propagation

alsuren committed:
one less unwrap

alsuren committed:
we don't just return paths

alsuren committed:
ConnectionStatus docstrings

alsuren committed:
ConnectionStatus::WatchdogTimeOut

qwandor-google commented on homie/Cargo.toml:
```diff
@@ -5,10 +5,10 @@ authors = ["Andrew Walbran <qwandor@google.com>", "David Laban <alsuren@gmail.co
+futures = "0.3.5"
+local_ipaddress = "0.1.3"
+log = "0.4.11"
+mac_address = "1.0.3"
```
What is the advantage of specifying an exact version like this, rather than just a compatible version?

alsuren replied:
This is what `cargo upgrade` does. 

I think the idea is that if someone already has a Cargo.lock with an older version of the dependency, and they install your lib, the version should automatically get bumped to one that's known to have all of the bug fixes etc. that your library has been tested against.

qwandor-google commented on run.sh:
```diff
@@ -9,22 +9,25 @@ TARGET=${TARGET:-armv7-unknown-linux-gnueabihf}
 elif [ $PROFILE = debug ]
 then
-    time cross build --target $TARGET
+    time cross build --target $TARGET --bin $BIN
```
I guess you need to specify --bin for the release build case above too.

qwandor-google commented on run.sh:
```diff
@@ -9,22 +9,25 @@ TARGET=${TARGET:-armv7-unknown-linux-gnueabihf}
 if [ $RUN -eq 1 ]
 then
-    ssh $TARGET_SSH ./publish-mqtt
+    ssh $TARGET_SSH sudo systemctl stop $BIN.service || echo "Oh. Nevermind."
```
Should this include the $SUFFIX?

alsuren replied:
no. we only have a single publish-mqtt.service installed. If you're trying to run something other than publish-mqtt then this command will fail and that's fine.

qwandor-google commented on publish-mqtt/src/main.rs:
```diff
@@ -90,29 +69,48 @@ async fn main() -> Result<(), Box<dyn Error + Send + Sync>> {
-        // LocalSet finished first. Colour me confused.
-        local.map(|()| Ok(println!("WTF?"))),
+    let res: Result<_, anyhow::Error> = try_join! {
+        // The resource is a task that should be spawned onto a tokio compatible
```
Is the 'resource' the `dbus_handle`?

alsuren replied:
yes. resource is just copy-paste from their example before I renamed it. Will fix.

qwandor-google commented on publish-mqtt/src/main.rs:
```diff
@@ -90,29 +69,48 @@ async fn main() -> Result<(), Box<dyn Error + Send + Sync>> {
         bluetooth_handle.map(|res| Ok(res?)),
         // MQTT event loop finished first.
-        mqtt_handle,
+        mqtt_handle.map(|res| Ok(res.map_err(|err| anyhow::anyhow!(err))?)),
```
Why do you need to do the `Ok(...?)` thing here? Isn't the error type already converted appropriately by the `map_err`?

alsuren replied:
I probably don't

alsuren committed:
stream closes when match is dropped

alsuren committed:
ConnectionStatus::MarkedDisconnected

alsuren committed:
set self.last_update_timestamp

alsuren committed:
handle readings from disconnected sensors

alsuren committed:
no need for Ok(...?) dance for mqtt_handle

alsuren committed:
stop scanning when we have all sensors

alsuren committed:
fix copy-pasted comment

alsuren committed:
allow specifying bin on release build

alsuren committed:
deploy and tail rather than run

qwandor-google commented on run.sh:
```diff
@@ -8,25 +8,32 @@ set -euxo pipefail
 
-if [ $RUN -eq 1 ]
-then
+if [ $DEPLOY = 1 ]; then
```
DEPLOY seems like a weird name for this setting, when it really means 'restart the service and follow it'. If I saw the name without reading the code I would guess that DEPLOY controlled whether it rsyncs the binary across.

qwandor-google commented on publish-mqtt/src/main.rs:
```diff
@@ -291,78 +383,87 @@ async fn disconnect_first_stale_sensor(
+                    "Got update from disconnected device {}. Connecting.",
+                    object_path
+                );
+                sensor.publish_readings(homie, &readings).await?;
```
You first need to call `homie.add_node`. You may also want to update the  `sensor.connection_status`.

qwandor-google commented on publish-mqtt/src/main.rs:
```diff
@@ -174,90 +173,153 @@ impl Sensor {
+            }
+            time::delay_for(CONNECT_INTERVAL).await;
+        }
+        #[allow(unreachable_code)]
```
This is kind of yucky, how would it look to declare the type of t1 instead?

(We should probably also give t1 and t2 better names.)

alsuren committed:
update status on unexpected readings

qwandor-google commented on publish-mqtt/src/main.rs:
```diff
@@ -121,30 +124,26 @@ impl Sensor {
         sensor_names: &HashMap<String, String>,
-    ) -> Result<Self, Box<dyn Error>> {
-        let mac_address = device.get_address()?;
+    ) -> Result<Self, anyhow::Error> {
```
This method can't actually fail, can it?

alsuren committed:
revert run.sh changes

They can come in their own PR

alsuren committed:
Sensor::new() can't fail

qwandor-google commented on publish-mqtt/src/main.rs:
```diff
@@ -174,90 +173,153 @@ impl Sensor {
     if let Some(mut sensor) = sensors_to_connect.pop_front() {
         println!("Trying to connect to {}", sensor.name);
-        match connect_start_sensor(bt_session, homie, &mut sensor).await {
+        match connect_start_sensor(bt_session.clone(), homie, &mut sensor).await {
```
This is an existing issue so it doesn't need to be fixed in this PR, but there are two different ways connect_start_sensor could fail: there could be an issue with the bluetooth connection, which the way it is handled below is fine for, or there could be an error sending the update to the `HomieDevice` (presumably due to the network connection being dropped), which for consistency should be bubbled up in the same way as disconnect_first_stale_sensor does.

alsuren replied:
Discussed in person, but I think it's worth writing down as well:

I think that this should be covered as part of https://github.com/alsuren/mijia-homie/issues/1 / https://github.com/alsuren/reading-xiaomi-temp/pull/32 (you would set connection_status to `::Connecting { reserved_until: five_minutes_time }` and then keep retrying the connect+subscribe until four_minutes_time and spend up to a minute cleaning up after yourself. The consumer of the connect_and_subscribe() function therefore only needs to know about the Connecting status (and not to try connecting while an attempt is in progress) and shouldn't need to care about any more of the details.

alsuren committed:
remove race condition comment

I've harped on about it enough, and I think I handle it
correctly in my less-lockey branch, so I can stop
worrying about it.

qwandor-google commented on publish-mqtt/src/main.rs:
```diff
@@ -272,17 +334,47 @@ async fn connect_first_sensor_in_queue(
+        .connect(&sensor.object_path)
+        .await
+        .with_context(|| std::line!().to_string())?;
+    match start_notify_sensor(bt_session.clone(), &sensor.object_path).await {
```
Could `start_notify_sensor` take a reference to bt_session instead, so you don't have to clone it?

qwandor-google commented on publish-mqtt/src/main.rs:
```diff
@@ -272,17 +334,47 @@ async fn connect_first_sensor_in_queue(
+            Ok(())
+        }
+        Err(e) => {
+            // If starting notifications failed, disconnect so that we start again from a clean
```
... failed a second time

alsuren committed:
extract bluetooth_connection_loop() function

alsuren committed:
stop cloning MijiaSession

alsuren committed:
update comment about failing twice

qwandor-google commented on mijia/src/lib.rs:
```diff
@@ -1,137 +1,107 @@
+pub const CONNECTION_INTERVAL_CHARACTERISTIC_PATH: &str = "/service0021/char0045";
 /// 500 in little-endian
-const CONNECTION_INTERVAL_500_MS: [u8; 3] = [0xF4, 0x01, 0x00];
+pub const CONNECTION_INTERVAL_500_MS: [u8; 3] = [0xF4, 0x01, 0x00];
```
Why do these constants all need to be pub?

alsuren replied:
they probably only needed to be pub while I was mid-refactor. Fixed now.

qwandor-google commented on mijia/src/lib.rs:
```diff
@@ -1,137 +1,107 @@
 
-const MIJIA_SERVICE_DATA_UUID: &str = "0000fe95-0000-1000-8000-00805f9b34fb";
+pub const MIJIA_SERVICE_DATA_UUID: &str = "0000fe95-0000-1000-8000-00805f9b34fb";
 pub const SERVICE_CHARACTERISTIC_PATH: &str = "/service0021/char0035";
```
This could do with a more descriptive name, maybe `SENSOR_READING_CHARACTERISTIC_PATH`.

alsuren committed:
constants don't need to be pub

qwandor-google commented on mijia/src/lib.rs:
```diff
@@ -1,137 +1,107 @@
+    bt_session: &MijiaSession,
+    device_path: &str,
+) -> Result<(), anyhow::Error> {
+    let temp_humidity_path: String = device_path.to_string() + SERVICE_CHARACTERISTIC_PATH;
```
Why do you need to explicitly say this is a String?

alsuren replied:
Doesn't need to be. Fixed.

alsuren committed:
SENSOR_READING_CHARACTERISTIC_PATH

qwandor-google commented on mijia/src/lib.rs:
```diff
@@ -1,137 +1,107 @@
+    let temp_humidity = dbus::nonblock::Proxy::new(
+        "org.bluez",
+        temp_humidity_path,
+        Duration::from_secs(30),
```
Put this as a constant up the top of the file. I assume it is a timeout of some sort?

qwandor-google commented on mijia/src/lib.rs:
```diff
@@ -1,137 +1,107 @@
+    let connection_interval = dbus::nonblock::Proxy::new(
+        "org.bluez",
+        connection_interval_path,
+        Duration::from_secs(30),
```
Ditto.

qwandor-google commented on mijia/src/session.rs:
```diff
@@ -0,0 +1,146 @@
+                let object_path = object_path
+                    .strip_suffix(crate::SERVICE_CHARACTERISTIC_PATH)?
+                    .to_owned();
+                let readings = crate::decode_value(&value)?;
```
Why not just use these symbols, rather than having crate:: all over the place?

alsuren replied:
mostly because it's easier when doing experimental refactoring, because you don't need to keep your imports up-to-date with your code manually. I will stop being lazy and fix them.

qwandor-google commented on mijia/src/session.rs:
```diff
@@ -0,0 +1,146 @@
+        dbus::nonblock::Proxy::new(
+            "org.bluez",
+            object_path.to_owned(),
+            Duration::from_secs(30),
```
This should be a constant too.

qwandor-google commented on publish-mqtt/src/main.rs:
```diff
@@ -291,78 +374,92 @@ async fn disconnect_first_stale_sensor(
+    println!("Subscribing to events");
+    let (msg_match, mut events) = bt_session.event_stream().await?;
+    println!("Processing events");
     // Process events until there are none available for the timeout.
```
Is this comment still accurate? If so, we should be retrying before panicking, as we might legitimately not get any events for a while.

alsuren replied:
Not accurate. Removed.

alsuren committed:
comment about stream never ending

alsuren committed:
remove pointless : String annotation

alsuren committed:
use crate::{...}

alsuren committed:
DBUS_METHOD_CALL_TIMEOUT constant

alsuren committed:
unused import

qwandor committed:
Merge branch 'master' into bluez-generated

alsuren committed:
oh yeah. I deleted read-all-devices
